### PR TITLE
python3Packages.torchvision: added cudaSupport option

### PR DESCRIPTION
**Pull request by yl3dy**
Added support for CUDA in torchvision, also updated pytorch to passthru
`cudaSupport` and `cudaArchList`. Torchvision also has a separate option
to turn off CUDA support.

###### Motivation for this change

Torchvision has an automatic detection of CUDA support in Pytorch via `torch.cuda.is_available()` call. This call always returns False in a sandbox, so it is needed to force CUDA build if Pytorch is CUDA-capable. This patch adds automatic passing of CUDA support flag and CUDA architecture list from Pytorch.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/torchvision/default.nix [hunk 1]
```diff
@@ -10,9 +11,19 @@
 , pillow
 , pytorch
 , pytest
+, cudatoolkit
+, cudnn
+, cudaSupport ? null  # by default uses the value from pytorch
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/torchvision/default.nix:16]:
```suggestion
, cudaSupport ? pytorch.cudaSupport or false # by default uses the value from pytorch
```

File: pkgs/development/python-modules/torchvision/default.nix [hunk 2]
```diff
@@ -10,9 +11,19 @@
...
 }:
 
-buildPythonPackage rec {
+let
+  cudatoolkit_joined = symlinkJoin {
+    name = "${cudatoolkit.name}-unsplit";
+    paths = [ cudatoolkit.out cudatoolkit.lib ];
+  };
+  final_cudaSupport = if cudaSupport == null then pytorch.cudaSupport else cudaSupport;
+  cudaArchStr = if final_cudaSupport then lib.strings.concatStringsSep ";" pytorch.cudaArchList else "";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/torchvision/default.nix:25]:
```suggestion
  cudaArchStr = lib.optionalString cudaSupport then lib.strings.concatStringsSep ";" pytorch.cudaArchList;
```

File: pkgs/development/python-modules/torchvision/default.nix [hunk 3]
```diff
@@ -23,15 +34,20 @@ buildPythonPackage rec {
     sha256 = "13j04ij0jmi58nhav1p69xrm8dg7jisg23268i3n6lnms37n02kc";
   };
 
-  nativeBuildInputs = [ libpng ninja which ];
+  nativeBuildInputs = [ libpng ninja which ] ++ lib.optional final_cudaSupport [ cudatoolkit_joined ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/torchvision/default.nix:37]:
```suggestion
  nativeBuildInputs = [ libpng ninja which ]
    ++ lib.optionals cudaSupport [ cudatoolkit_joined ];
```

File: pkgs/development/python-modules/torchvision/default.nix [hunk 4]
```diff
@@ -23,15 +34,20 @@ buildPythonPackage rec {
...
 
   TORCHVISION_INCLUDE = "${libjpeg_turbo.dev}/include/";
   TORCHVISION_LIBRARY = "${libjpeg_turbo}/lib/";
 
-  buildInputs = [ libjpeg_turbo libpng ];
+  buildInputs = [ libjpeg_turbo libpng ] ++ lib.optional final_cudaSupport [ cudnn ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/torchvision/default.nix:42]:
```suggestion
  buildInputs = [ libjpeg_turbo libpng ]
    ++ lib.optionals cudaSupport [ cudnn ];
```

File: pkgs/development/python-modules/torchvision/default.nix [hunk 5]
```diff
@@ -23,15 +34,20 @@ buildPythonPackage rec {
...
 
   propagatedBuildInputs = [ numpy pillow pytorch scipy ];
 
+  preBuild = if final_cudaSupport then ''
+    export TORCH_CUDA_ARCH_LIST="${cudaArchStr}"
+    export FORCE_CUDA=1
+  '' else "";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/torchvision/default.nix:49]:
```suggestion
  preBuild = lib.optionalString final_cudaSupport ''
    export TORCH_CUDA_ARCH_LIST="${cudaArchStr}"
    export FORCE_CUDA=1
  '';
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Review by yl3dy — COMMENTED** [review]
@SuperSandro2000 Thanks for suggestions!

**Pull request merged**